### PR TITLE
add normalizeQueryKeepNames and normalizedQueryHashKeepNames

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add `normalizeQueryKeepNames` and `normalizedQueryHashKeepNames` to normalize queries without masking long names with `?`. This helps better analyze complex query logs.

Detailed description / Documentation draft:

.

### CODE REVIEW

**Comment by amosbird** [comment]
ANTLR parser doesn't seem to work with column transformers

https://clickhouse-test-reports.s3.yandex.net/18910/90e9f4fc2e33098ad2442f53758489fc697ce431/functional_stateless_tests_(antlr_debug)/test_run.txt.out.log

**Comment by alexey-milovidov** [comment]
ANTLR parser is not yet complete.
You can add an exception to skip_list.json

**Comment by alexey-milovidov** [comment]
I have a feeling that `WithNames` meaning is non-obvious, it's not clear whether it is
- normalizing names as well (1);
- keeping names (2).

(in fact it is 2)

**Comment by amosbird** [comment]
> I have a feeling that `WithNames` meaning is non-obvious, it's not clear whether it is
> 
> * normalizing names as well (1);
> * keeping names (2).
> 
> (in fact it is 2)

Yeah, let's switch to `-KeepNames`

**Comment by alexey-milovidov** [comment]
Ok.

File: src/Parsers/queryNormalization.h [hunk 1]
```diff
@@ -10,6 +10,7 @@
 
 namespace DB
 {
+template <bool with_names>
```
Comment (Akazz) [inline src/Parsers/queryNormalization.h:13]: I'd suggest you rename template parameter here (and in the method below) as well

File: src/Parsers/queryNormalization.h [hunk 2]
```diff
@@ -61,31 +62,39 @@ inline UInt64 ALWAYS_INLINE normalizedQueryHash(const char * begin, const char *
             /// By the way, there is padding in columns and pointer dereference is Ok.
             || (token.type == TokenType::BareWord && *token.end != '('))
         {
-            /// Identifier is complex if it contains whitespace or more than two digits
-            /// or it's at least 36 bytes long (UUID for example).
-            size_t num_digits = 0;
-
-            const char * pos = token.begin;
-            if (token.size() < 36)
+            /// Explicitly ask to normalize with identifier names
```
Comment (Akazz) [inline src/Parsers/queryNormalization.h:65]: It probably should be "... to normalize keeping/retaining identifier names".

**Review by Akazz — COMMENTED** [review]
Looking good in the rest of it

File: src/Parsers/queryNormalization.h (hunk 1 above)
Comment (amosbird) [reply src/Parsers/queryNormalization.h:13]: Done

File: src/Parsers/queryNormalization.h (hunk 2 above)
Comment (amosbird) [reply src/Parsers/queryNormalization.h]: Done

**Comment by ramazanpolat** [comment]
@amosbird 

Can you please explain what the difference is between `normalizedQueryHash` and `normalizedQueryHashKeepNames`?

They both generate `UInt64` hash values and they produce same results most of the time.

**Comment by amosbird** [comment]
@ramazanpolat `normalizeQuery` will replace identifier with `?` if it contains whitespace or more than two digits
 or it's at least 36 bytes long (UUID for example).
 `normalizeQueryWithNames` will keep identifier names intact. `-Hash` is based on this.

**Pull request merged**